#### Promise(一)

> Promise的定义：
>
> Promise是用来解决异步代码中回调地狱的问题而诞生的产物

#### 回调地狱

> 在开发中，通常会出现回调函数中嵌套回调的情况，如果回调函数过多，就会出现回调地狱，导致代码可读性严重下降，例如下面这样：

```javascript
setTimeout(function(){
    console.log('1')
    setTimeout(function(){
    	console.log('2')
        setTimeout(function(){
    		console.log('3')
            setTimeout(function(){
    			console.log('4')
                setTimeout(function(){
    				console.log('5')
				},2000)
			},2000)
		},2000)
	},2000)
},2000)
```

> 上面这段代码的执行效果是，
>
> 两秒中之后在控制台输出1
>
> 过两秒中之后在控制台输出2
>
> 在过两秒中之后在控制台输出3
>
> 然后再过两秒中之后在控制台输出4
>
> 最后在过两秒中在控制台输出5

> 这种回调函数嵌套回调函数的写法就是回调地狱
>
> 为了让代码的可读性更加的高，
>
> javascript的作者 Brendan Eich开发出了Promise这个类
>
> 基础写法如下：

```javascript
new Promise(function(resolve,reject){
    //在这里写异步方法；比如 ajax,xhr,setTimeout等，
    //还可以写任何的非异步的回调函数；
    //但在开发中，大多数情况下，是为处理异步而生；
    ajax({
        url:'/user/info',
        success:res => resolve(res),
        fail:err => reject(err)
    })
    //通常情况下，异步代码执行成功时会调用 resolve(res) 方法；
    //异步代码调用失败时会走 reject(err) 方法；
})
```

####  then和catch

> 在Promise的实例中，有两个方法，一个是then方法，另一个 catch方法，
>
> 他们的使用如下所示：

```javascript
//then方法的使用方式；
new Promise().then(res=>{
    console.log(res);
})
//catch方法的使用方式
new Promise().catch(err=>{
    console.log(err);
})
```

#### then、catch和resolve、reject的关系

> 你现在应该已经会书写then方法及catch方法了，
>
> 但此时你应该看不懂 .then中的res及.catch中的err到底是个什么意思
>
> 当你在控制台输出时，发现他们根本就没有走这两个方法；
>
> 因为在Promise中，只有执行了resolve()方法，then方法才会执行
>
> 当执行了reject()方法时才会执行catch()方法；
>
> 例如这样：

```javascript
// .then方法和resolve()方法的使用：
new Promise((resolve,reject)=>{
    resolve('hello world!')
}).then(res=>{
    console.log(res) // 注：res的值是resolve方法中的实参 ： 'hello world!'
})

// .catch方法和reject()方法的使用：
new Promise((resolve,reject)=>{
    reject('is typeError')
}).catch(err=>{
    console.log(err) // 注：err的值是reject方法中的实参 ： 'is typeError'
})
```

> 扩展：.then 方法和.catch方法是支持链式调用的，比如这样：

```javascript
new Promise().then(res=>{}).then(res=>{}).then(res=>{}).catch(err=>{}).then(res=>{})
```

> Tip:
>
> 如果一个Promise中有多个.then方法那么会顺序的调用then()方法并执行，如果有一个方法报错，则直接走catch方法，报错后即使还有then方法没有执行，也不会在执行了，效果类似try catch中的catch

> 到此，我们已经学会了Promise的基本使用，我们可以把在文章开始时的地狱回调问题做一个优化，第一步,按照Promise的定义：把异步回调添加到Promise中，2秒钟之后走resolve()方法，然后通过then方法接收，代码如下：

```javascript
new Promise(function(resolve,reject){
    setTimeout(function(){
        resolve()
    },2000)
})
.then(res=>{
    console.log(1)
})
```

> 上面代码中，已经是很规范的书写方式，但因为Promise本身也是类,所以更多情况下可以用实例的接收方式来调用，这样显得代码可读性更高，比如这样：

```javascirpt
let p = new Promise(function(resolve,reject){
    setTimeout(function(){
        resolve()
    },2000)
})

p.then(res=>{
    console.log(1)
})
```

> 我们的题是当第一个函数执行完毕后控制台输出一个2，那么我们的代码会变成这样

```javascript
let p = new Promise(function(resolve,reject){
    setTimeout(function(){
        resolve()
    },2000)
})

p.then(res=>{
    console.log(1);
    new Promise(function(resolve,reject){
    	setTimeout(function(){
        	resolve()
    	},2000)
	})
    .then(res={
        console.log(2);
    })
})
```

> 这样就完成了在2秒钟之后输出1，在停2秒输出2，但是代码并没有解决地狱回调的问题，而是在then方法中不断的写回调，此时我们需要创建一个fn方法+return Promise的方式来解决这个问题，代码变成了这样：

```javascript
let fn = function(){
    return new Promise(function(resolve,reject){
    	setTimeout(function(){
        	resolve()
    	},2000)
	})
}
let p = fn();

p.then(res=>{
    console.log(1);
    return fn()
})

.then(res=>{
    console.log(2);
})
```

> 闭包的使用:
>
> 我们现在已经利用实例对象+fn方法+then方法解决了地狱回调的问题，
>
> 接着，我们可以再优化一下fn的方法，给fn方法加一个入参：

```javascript
//原理是： 把fn的入参传给resolve，因为.then方法会接收到resolve的值，所以.then中res的值就是我们想要的控制台输出结果，这其中用到的知识，就是闭包
let fn = function(num){
    return new Promise(function(resolve,reject){
    	setTimeout(function(){
        	resolve(num)
    	},2000)
	})
}

fn(1).then(res=>{
    console.log(res);
    return fn(2)
})

.then(res=>{
    console.log(res);
    return fn(3);
})
```

> 接着我们在多写几个promise，就几乎完工了

```javascript
let fn = function(num){
    return new Promise(function(resolve,reject){
        setTimeout(function(){
            resolve(num)
        },2000)
    })
}

fn(1).then(res=>{
    console.log(res);
    return fn(2)
})

.then(res=>{
    console.log(res);
    return fn(3);
})

.then(res=>{
    console.log(res);
    return fn(4)
})

.then(res=>{
    console.log(res)
    return fn(5)
})
```

> 对比最上面的地狱回调，我们已经实现了代码解耦，可读性变得很高
>
> 但是这并不是最优的解，观察这道题，我们可以promise中的链式调用，在fn方法中利用 then方法，直接输出我们想要的结果，代码如下：

```javascript
let fn = function(num){
    return new Promise(function(resolve,reject){
        setTimeout(function(){
            resolve(num)
        },2000)
    }).then(res=>{
        console.log(num);
    })
}

fn(1).then(res=>fn(2)).then(res=>fn(3)).then(res=>fn(4)).then(res=> fn(5))
```

